File: gtci/17-modified-binary-search/cpp/4-number-range.cpp
#include <cassert>
#include <iostream>
#include <vector>

class Solution {
public:
  std::vector<int> findRange(const std::vector<int>& arr, int key) {
    return {findFirstOccurrence(arr, key), findLastOccurrence(arr, key)};
  }
private:
  int findFirstOccurrence(const std::vector<int>& arr, const int& key) {
    int left = 0;
    int right = arr.size() - 1;

    int first = -1;

    while (left <= right) {
      const int mid = left + (right - left) / 2;

      if (arr[mid] == key) {
        first = mid;
        right = mid - 1;
      } else if (arr[mid] < key) {
        left = mid + 1;
      } else if (arr[mid] > key) {
        right = mid - 1;
      }
    }

    return first;
  }

  int findLastOccurrence(const std::vector<int>& arr, const int& key) {
    int left = 0;
    int right = arr.size() - 1;

    int last = -1;

    while (left <= right) {
      const int mid = left + (right - left) / 2;

      if (arr[mid] == key) {
        last = mid;
        left = mid + 1;
      } else if (arr[mid] < key) {
        left = mid + 1;
      } else if (arr[mid] > key) {
        right = mid - 1;
      }
    }

    return last;
  }
};

int main() {
    Solution s;

    std::vector<int> arr1 {4, 6, 6, 6, 9};
    int key1 = 6;
    std::vector<int> result1 = s.findRange(arr1, key1);
    assert((result1 == std::vector<int>{1, 3}));

    std::vector<int> arr2 {1, 3, 8, 10, 15};
    int key2 = 10;
    std::vector<int> result2 = s.findRange(arr2, key2);
    assert((result2 == std::vector<int>{3, 3}));

    std::vector<int> arr3 {1, 3, 8, 10, 15};
    int key3 = 12;
    std::vector<int> result3 = s.findRange(arr3, key3);
    assert((result3 == std::vector<int>{-1, -1}));

    std::cout << "All test cases passed." << std::endl;

    return 0;
}

File: gtci/17-modified-binary-search/python3/4-number-range.py
class Solution:
    def findRange(self, arr, key):
        if not arr:
            return [-1, -1]

        def find_first_occurrence():
            left = 0
            right = len(arr) - 1
            first = -1

            while left <= right:
                mid = (left + right) // 2

                if arr[mid] == key:
                    first = mid
                    right = mid - 1
                elif arr[mid] < key:
                    left = mid + 1
                elif arr[mid] > key:
                    right = mid - 1

            return first

        def find_last_occurrence():
            left = 0
            right = len(arr) - 1
            last = -1

            while left <= right:
                mid = (left + right) // 2

                if arr[mid] == key:
                    last = mid
                    left = mid + 1
                elif arr[mid] < key:
                    left = mid + 1
                elif arr[mid] > key:
                    right = mid - 1

            return last

        return [find_first_occurrence(), find_last_occurrence()]


if __name__ == "__main__":
    sol = Solution()

    assert sol.findRange([4, 6, 6, 6, 9], 6) == [1, 3]
    assert sol.findRange([1, 3, 8, 10, 15], 10) == [3, 3]
    assert sol.findRange([1, 3, 8, 10, 15], 12) == [-1, -1]

    print("All test cases passed.")
